sink.js

```diff
@@ -1,112 +1,93 @@
+  }
 
-      // .end() was called, waiting on flush (finish event)
-      if (result.ending) {
-        await Promise.race([
-          errPromise.promise,
-          // TODO: do we need to wait on close? If slow end and destroy is
-          // called then what is emitted? close or finish?
-          closePromise.promise,
-          finishPromise.promise
-        ])
-        break
-      }
+  const waitForDrainOrClose = () => {
+    return new Promise((resolve, reject) => {
+      closeCb = drainCb = resolve
+      errCb = reject
+      writable.once('drain', drainHandler)
+    })
+  }
 
-      // If destroyed, race err & close to determine reason & then throw/break
-      if (writable.destroyed) {
-        await Promise.race([
-          errPromise.promise,
-          closePromise.promise
-        ])
-        break
-      }
+  const waitForDone = () => {
+    // Immediately end the source
+    endSource(source)
+    return new Promise((resolve, reject) => {
+      if (closed || finished) return resolve()
+      finishCb = closeCb = resolve
+      errCb = reject
+    })
+  }
 
-      if (result.done) {
-        writable.end()
-        await Promise.race([
-          errPromise.promise,
-          // TODO: do we need to wait on close? If slow end and destroy is
-          // called then what is emitted? close or finish?
-          closePromise.promise,
-          finishPromise.promise
-        ])
-        break
-      }
+  const cleanup = () => {
+    writable.removeListener('error', errorHandler)
+    writable.removeListener('close', closeHandler)
+    writable.removeListener('finish', finishHandler)
+    writable.removeListener('drain', drainHandler)
+  }
+
+  writable.once('error', errorHandler)
+  writable.once('close', closeHandler)
+  writable.once('finish', finishHandler)
 
-      if (!writable.write(result.value)) {
-        drainPromise = defer()
-        await Promise.race([
-          errPromise.promise,
-          closePromise.promise,
-          finishPromise.promise,
-          drainPromise.promise
-        ])
+  try {
+    for await (const value of source) {
+      if (!writable.writable || writable.destroyed) break
+
+      if (writable.write(value) === false) {
+        await waitForDrainOrClose()
       }
     }
-  } finally {
-    writable
-      .removeListener('error', errorHandler)
-      .removeListener('close', closeHandler)
-      .removeListener('finish', finishHandler)
-      .removeListener('drain', drainHandler)
-
-    // End the iterator if it is a generator
-    if (typeof source.return === 'function') {
-      await source.return()
-    }
+  } catch (err) {
+    // The writable did not error, give it the error
+    writable.destroy(err)
   }
+
+  // Everything is good and we're done writing, end everything
+  if (!error && writable.writable) {
+    writable.end()
+  }
+
+  // Wait until we close or finish. This supports halfClosed streams
+  await waitForDone()
+
+  // Clean up listeners
+  cleanup()
```

Comment:
Should this me be more like:

```javascript
try {
  await waitForDone()
} finally {
  // Clean up listeners
  cleanup()
}
```

`waitForDone` assigns `reject` to `errCb` which is invoked in `errorHandler` which will be invoked in turn if there's an `error` event on the writable so we probably still want to remove the listeners?

Reply:
yes, will update

Reply:
I've also added writable.end() to the try catch. It **shouldnt** throw, but in the event it does, this will ensure the handlers get cleaned up.